File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.js
function solution(s) {
    var answer = 0;
    let dic = {'(' : 0, ')' : 0, '[' : 0, ']' : 0, '{' : 0, '}' : 0}
    for (let i = 0; i < s.length; i++) {
        dic[s[i]] += 1
    }
    if (dic['('] !== dic[')'] || dic['{'] !== dic['}'] || dic['['] !== dic[']']) {
        return answer
    }
    for (let i = 0; i < s.length; i++) {
        let stack = []
        let tmp = s.slice(i) + s.slice(0,i)
        for (let idx = 0; idx < tmp.length; idx++) {
            if (tmp[0] === ')' || tmp[0] === ']' || tmp[0] === '}') {
                continue
            }
            else {
                if (idx === 0) {
                    stack.push(tmp[idx])
                } else {
                    if (tmp[idx] === '}' && stack[stack.length-1] === '{') {
                        stack.pop()
                    }
                    else if (tmp[idx] === ']' && stack[stack.length-1] === '[') {
                        stack.pop()
                    }
                    else if (tmp[idx] === ')' && stack[stack.length-1] === '(') {
                        stack.pop()
                    }
                    else {
                        stack.push(tmp[idx])
                    }
                }
            }
            // console.log(stack)
            // console.log('안임')
            if (idx === tmp.length - 1 && stack.length === 0) {
                answer += 1
            } 
        }
        // console.log(stack)
        // console.log('밖임')
        // if (stack.length === 0) {
        //     answer += 1
        // }
    }
    return answer;
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
def solution(s):
    answer = 0
    stack = []
    dict = {'(' : 0, ')':0, '[' : 0, ']' : 0, '{' : 0, '}' : 0}
    for char in s:
        dict[char] += 1
    if dict['('] != dict[')'] or dict['{'] != dict['}'] or dict['['] != dict[']']:
        #애초에 닫는 괄호, 여는 괄호의 수가 다르다면 result 올릴 수 없다
        return 0
    for i in range(len(s)):
        flag = False
        newS = s[i:] + s[:i]
        for char in newS:
            if char == '(' or char == '{' or char == '[':
                stack.append(char)
                flag = True
            else:
                if not stack:
                    flag = False
                    break
                else:
                    if char == ')' and stack[-1] == '(':
                        stack.pop()
                    elif char == '}' and stack[-1] == '{':
                        stack.pop()
                    elif char == ']' and stack[-1] == '[':
                        stack.pop()
                    else:
                        break
        if flag and not stack:
            answer += 1

    return answer